elements/pl-checkbox/pl-checkbox.py

```diff
@@ -152,14 +154,17 @@ def render(element_html, data):
     if data['panel'] == 'question':
         partial_score = data['partial_scores'].get(name, {'score': None})
         score = partial_score.get('score', None)
+        feedback = partial_score.get('feedback', None)
 
         answerset = []
         for answer in display_answers:
             answer_html = {
                 'key': answer['key'],
                 'checked': (answer['key'] in submitted_keys),
                 'html': answer['html'].strip(),
-                'display_score_badge': score is not None and show_answer_feedback and answer['key'] in submitted_keys
+                'display_score_badge': score is not None and show_answer_feedback and answer['key'] in submitted_keys,
+                'display_feedback': feedback and answer['feedback'] in feedback and feedback[0] is not None,
+                'feedback': feedback.pop(0) if feedback and answer['feedback'] in feedback else None
```

Comment:
I'm a little confused with what's going on here:

- I don't understand what `answer['feedback'] in feedback` and `feedback[0] is not None` are checking for, and specifically why they're referencing the `feedback` variable - it seems like it would be sufficient to just write `answer.get('feedback', None) is not None`.
- Similarly for `feedback.pop(0) if feedback and answer['feedback'] in feedback else None`, I suspect that could just be written as `answer.get('feedback', None)`.

Regardless of all that, we need to avoid unchecked access to `answer['feedback']` for backwards compatibility with existing question instances (e.g., trying to render an existing question fails with `KeyError`).

Reply:
Yeah I could have made this more clear, my bad. Allow me to explain:

- `feedback` is an attribute of the `partial scores` object. In `grade()` it gets populated with the feedback for each answer choice that was selected by the student, in order. For example, if  A, C, and E are chosen, `feedback` would look like `[A's feedback, C's feedback, E's feedback]`. `answer['feedback']` is the feedback attached to the answer itself.

To answer your questions:

- `answer['feedback'] in feedback` is meant to check if the answer choice was chosen by the student. Looking at this again I suppose it would be better to replace it with `answer['key'] in submitted_keys`. In the example in the previous section, if A didn't have feedback, `A's feedback` would be `None`. `feedback[0] is not None` checks that if a student submitted one of these answers that don't contain feedback, it doesn't get displayed.  Otherwise, an empty gray box would appear.
- `feedback.pop(0) if feedback and answer['feedback'] in feedback else None` gets the actual feedback to display from the `feedback` list. Using the same reasoning as before, I suppose I can replace `answer['feedback'] in feedback` with `answer['key'] in submitted_keys`. 

With these edits, there are no more accesses to `answer['feedback']`. I'm hesitant to use `answer.get('feedback', None)` so I can be consistent with `pl-multiple-choice`, where I was told to put `feedback` into `partial_scores` and use it in `render()` (https://github.com/PrairieLearn/PrairieLearn/pull/4656#issuecomment-899773221). I'll have the changes pushed shortly, let me know what you think :)